concourse/steps/release.py

```diff
@@ -23,22 +29,317 @@
+        self.release_version = not_empty(release_version)
+        self.component_descriptor_file_path = existing_file(component_descriptor_file_path)
+
+    def name(self):
+        return "Release Step"
+
+    def validate(self):
+        info(f'validating {self.name()}')
+        # fetch release notes and generate markdown to catch errors early
+        release_notes = fetch_release_notes(
+            git_helper=self.git_helper,
+            github_helper=self.github_helper,
+            repository_branch=self.githubrepobranch.branch(),
+        )
+        release_nots.to_markdown()
+        self.release_context.set_release_notes(release_notes)
+
+        if self.component_descriptor_file_path:
+            # TODO: proper validation
+            existing_file(self.component_descriptor_file_path)
+            with open(self.component_descriptor_file_path) as f:
+                self.component_descriptor_contents = f.read()
+
+    def apply(
+        self,
+    ):
+        info('Create release')
+        # Create GitHub-release
+        release = self.github_helper.create_release(
+            tag_name=self.release_version,
+            body=self.release_context.release_notes().to_markdown(),
+            draft=False,
+            prerelease=False,
+        )
+
+        # Upload component descriptor to GitHub-release if one has been calculated
+        if self.component_descriptor_file_path:
+            release.upload_asset(
+                content_type='application/x-yaml',
+                name=product.model.COMPONENT_DESCRIPTOR_ASSET_NAME,
+                asset=self.component_descriptor_contents,
+                label=product.model.COMPONENT_DESCRIPTOR_ASSET_NAME,
+            )
+
+    def revert(self):
+        #TODO: remove release from GitHub
+        pass
+
+
+class DevCyclePrepareStep(TransactionalStep):
+    def __init__(
+        self,
+        github_helper: GitHubRepositoryHelper,
+        repo_dir: str,
+        repository_version_file_path: str,
+        release_version: str,
+        version_operation: str,
+        prerelease_suffix: str,
+        release_context: ReleaseContext,
+    ):
+        super().__init__(release_context)
+        self.github_helper = not_none(github_helper)
+        self.repo_dir=repo_dir
+        self.repository_version_file_path = repository_version_file_path
+
+        self.release_version = not_empty(release_version)
+        self.version_operation = not_empty(version_operation)
+        self.prerelease_suffix = not_empty(prerelease_suffix)
+
+    def name(self):
+        return "Next Cycle Commit Step"
+
+    def validate(self):
+        info(f'validating {self.name()}')
+        existing_file(pathlib.Path(self.repo_dir, self.repository_version_file_path))
+        next_cycle_version = version.process_version(
+            version_str=self.release_version,
+            operation=self.version_operation,
+        )
+        self.next_cycle_dev_version = version.process_version(
```

Comment:
actually, I meant to place the "this is superbad" comment here. but the diff was updated before i could save. so as mentioned earlier: NEVER modify state during validation